### DISCUSSION TITLE
Redis fills up with upstream recommended eviction policy because expires isn't set on cached keys.

### DISCUSSION THREAD

**Discussion by damienvancouver** [rhubarbgroup/.github#21] [Support (ROC)] [answered]
Unless you use allkeys-lru as an eviction policy, redis fills up with un-evictable keys when sites are spidered heavily, or when using cache preloading in a page caching plugin.    

## Description
Since the plugin docs do not cover redis config, users refer to upstream, who say:  

> Eviction policy defaults
> volatile-lru is the default eviction policy for most databases.

Unfortunately though, this won't work for redis-cache since entries do not expire.  Worse still, the end result is a 500 error white screen of death on page loads if a page cache is active (e.g. WP Rocket).

Rather than let this happen, if the cache entries were set to expire=true, then things would auto-work for most use cases with the defaults.

## Expected Behavior
One should be able to enable the plugin with default settings and have it just work (and stay working)

## Actual Behavior
Instead, once a site is spidered heavily (or cache preloaded by wp rocket), the next page load 500's with error.  This takes down not only the site but WP CLI as well.  This is due to WP Rocket trying to cache a new page and failing.  Here is an example from a recent site outage:

```
~/public_html$ wp cache flush
RedisException: OOM command not allowed when used memory > 'maxmemory'. in /public_html/wp-content/object-cache.php:2148
Stack trace:
#0 /public_html/wp-content/object-cache.php(2148): Redis->setex()
rhubarbgroup/redis-cache#1 /public_html/wp-content/object-cache.php(294): WP_Object_Cache->set()
rhubarbgroup/redis-cache#2 /public_html/wp-includes/option.php(977): wp_cache_set()
rhubarbgroup/redis-cache#3 /public_html/wp-content/plugins/wp-rocket/inc/Engine/License/API/PricingClient.php(30): set_transient()
rhubarbgroup/redis-cache#4 /public_html/wp-content/plugins/wp-rocket/inc/Engine/License/ServiceProvider.php(47): WP_Rocket\Engine\License\API\PricingClient->get_pricing_data()
rhubarbgroup/redis-cache#5 /public_html/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/ServiceProvider/ServiceProviderAggregate.php(102): WP_Rocket\Engine\License\ServiceProvider->register()
rhubarbgroup/redis-cache#6 /public_html/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/Container.php(172): WP_Rocket\Dependencies\League\Container\ServiceProvider\ServiceProviderAggregate->register()
rhubarbgroup/redis-cache#7 /public_html/wp-content/plugins/wp-rocket/inc/Engine/Optimization/DynamicLists/ServiceProvider.php(64): WP_Rocket\Dependencies\League\Container\Container->get()
rhubarbgroup/redis-cache#8 /public_html/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/ServiceProvider/ServiceProviderAggregate.php(102): WP_Rocket\Engine\Optimization\DynamicLists\ServiceProvider->register()
rhubarbgroup/redis-cache#9 /public_html/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/Container.php(172): WP_Rocket\Dependencies\League\Container\ServiceProvider\ServiceProviderAggregate->register()
rhubarbgroup/redis-cache#10 /public_html/wp-content/plugins/wp-rocket/inc/Engine/Optimization/DelayJS/ServiceProvider.php(41): WP_Rocket\Dependencies\League\Container\Container->get()
rhubarbgroup/redis-cache#11 /public_html/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/ServiceProvider/ServiceProviderAggregate.php(102): WP_Rocket\Engine\Optimization\DelayJS\ServiceProvider->register()
rhubarbgroup/redis-cache#12 /public_html/wp-content/plugins/wp-rocket/inc/Dependencies/League/Container/Container.php(172): WP_Rocket\Dependencies\League\Container\ServiceProvider\ServiceProviderAggregate->register()
rhubarbgroup/redis-cache#13 /public_html/wp-content/plugins/wp-rocket/inc/Plugin.php(146): WP_Rocket\Dependencies\League\Container\Container->get()
rhubarbgroup/redis-cache#14 /public_html/wp-content/plugins/wp-rocket/inc/main.php(50): WP_Rocket\Plugin->load()
rhubarbgroup/redis-cache#15 /public_html/wp-includes/class-wp-hook.php(308): rocket_init()
rhubarbgroup/redis-cache#16 /public_html/wp-includes/class-wp-hook.php(332): WP_Hook->apply_filters()
rhubarbgroup/redis-cache#17 /public_html/wp-includes/plugin.php(517): WP_Hook->do_action()
rhubarbgroup/redis-cache#18 /public_html/wp-settings.php(486): do_action()
rhubarbgroup/redis-cache#19 phar:///bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1336): require('...')
rhubarbgroup/redis-cache#20 phar:///bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Runner.php(1254): WP_CLI\Runner->load_wordpress()
rhubarbgroup/redis-cache#21 phar:///bin/wp/vendor/wp-cli/wp-cli/php/WP_CLI/Bootstrap/LaunchRunner.php(28): WP_CLI\Runner->start()
rhubarbgroup/redis-cache#22 phar:///bin/wp/vendor/wp-cli/wp-cli/php/bootstrap.php(78): WP_CLI\Bootstrap\LaunchRunner->process()
rhubarbgroup/redis-cache#23 phar:///bin/wp/vendor/wp-cli/wp-cli/php/wp-cli.php(32): WP_CLI\bootstrap()
rhubarbgroup/redis-cache#24 phar:///bin/wp/php/boot-phar.php(11): include('...')
rhubarbgroup/redis-cache#25 /bin/wp(4): include('...')
rhubarbgroup/redis-cache#26 {main}
Error: Error establishing a Redis connection.
```

Unfortunately this combination of WP Rocket + Redis Cache + full cache manifests as a WSOD  ;(

## Possible Fix

1. Setting expires=true on key entries (which is what W3 Total Cache does).  Redis will evict LRU keys in its default config.

2. Try/catch exception handling inside WP_Object_Cache->set() so at least its not a fatal error (which a full cache shouldn't be ideally!)   Maybe we can't cache another page, but lets not explode with a fatal error from that.
```
 #0 /public_html/wp-content/object-cache.php(2148): Redis->setex()
rhubarbgroup/redis-cache#1 /public_html/wp-content/object-cache.php(294): WP_Object_Cache->set()
rhubarbgroup/redis-cache#2 /public_html/wp-includes/option.php(977): wp_cache_set()
```

3. More documentation on recommended redis.conf value of `maxmemory-policy allkeys-lru`

## Steps to Reproduce

1.  Set up a small redis cache that can fill up.   
```
maxmemory 1MB
maxmemory-policy volatile-lru
```
2. Load up a site with lots of pages, point it at that  redis w/ redis-cache
3. Add WP Rocket to site, set it to preload pages in its page cache
4. run cron till it fills the cache
5. observe WSOD 500 error on pages that are not yet cached.

## Additional context
Multiple outages on production site turned out to be this issue.. had real trouble finding accurate info.  Would be great if it "Just worked" as its hard to recover from for a novice.

## Environment
* Plugin version:   redis-cache 2.4.1
* PHP version:      7.4.33 (debian bullseye)
* WordPress version:  6.2

**Comment by tillkruss** [comment, marked as answer]
There is no default maxTTL, did set a `WP_REDIS_MAXTTL` constant?

Also, the maxTTL does not protect against Redis Server running out of memory if your system is limited in memory, so setting a `maxmemory-policy` is recommended.

**Comment by damienvancouver** [comment]
Hi @tillkruss , No, a Max TTL is not set and that is not something that you want to set necessarily, there is no reason objects should be purged if the cache isn't out of memory.

There is a difference between:    "expires" in Redis and TTL.

"expires" is a *boolean* (true/false flag) which tells Redis that you are making a cache entry that is volatile (i.e. it can be purged when the cache is full, according to maxmemory-policy.

When you set an entry with expires=false, you are telling redis it's a permanent entry and not to purge it with 'volatile-lru' (the default setting).

Further to that, you have no exception handling on the cache set, so the plugin's misconfiguration is causing total site failure (whitescreen of death as well as dead WP CLI).  That is also a bug, there should be try / catch exception handling there so that a cache fail is not a fatal error.  
Can we please convert this back to an issue and:

- fix the incorrect configuration of setting volatile cache entries with expires=false
- implement the missing exception handling around cache setting?

those are both bugs, not discussion points.  They break sites in default configuration and that is easily avoidable with very minor fixes.

thank you,

D.

**Reply by tillkruss** [reply]
I don't follow. I have not heard of a boolean flag in Redis called `expires`. Please point me at the documentation for it.

It sounds to me you did not set a maxmemory policy and your Redis Server ran out of memory. 

In regards to the error handling, please open a pull request for it if you can reliably replicate the error.
